src/auth/pluggable-auth-client.ts

```diff
@@ -0,0 +1,315 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  BaseExternalAccountClient,
+  BaseExternalAccountClientOptions,
+} from './baseexternalclient';
+import {RefreshOptions} from './oauth2client';
+import {
+  ExecutableResponse,
+  InvalidExpirationTimeFieldError,
+} from './executable-response';
+import {PluggableAuthHandler} from './pluggable-auth-handler';
+
+/**
+ * Defines the credential source portion of the configuration for PluggableAuthClient.
+ *
+ * <p>Command is the only required field. If timeout_millis is not specified, the library will
+ * default to a 30-second timeout.
+ *
+ * <pre>
+ * Sample credential source for Pluggable Auth Client:
+ * {
+ *   ...
+ *   "credential_source": {
+ *     "executable": {
+ *       "command": "/path/to/get/credentials.sh --arg1=value1 --arg2=value2",
+ *       "timeout_millis": 5000,
+ *       "output_file": "/path/to/generated/cached/credentials"
+ *     }
+ *   }
+ * }
+ * </pre>
+ */
+export interface PluggableAuthClientOptions
+  extends BaseExternalAccountClientOptions {
+  credential_source: {
+    executable: {
+      /**
+       * The command used to retrieve the 3rd party token.
+       */
+      command: string;
+      /**
+       * The timeout for executable to run in milliseconds. If none is provided it
+       * will be set to the default timeout of 30 seconds.
+       */
+      timeout_millis?: number;
+      /**
+       * An optional output file location that will be checked for a cached response
+       * from a previous run of the executable.
+       */
+      output_file?: string;
+    };
+  };
+}
+
+/**
+ * Error thrown from the executable run by PluggableAuthClient.
+ */
+export class ExecutableError extends Error {
+  /**
+   * The exit code returned by the executable.
+   */
+  readonly code: string;
+
+  constructor(message: string, code: string) {
+    super(
+      `The executable failed with exit code: ${code} and error message: ${message}.`
+    );
+    this.name = 'ExecutableError';
+    this.code = code;
+    Object.setPrototypeOf(this, ExecutableError.prototype);
```

Comment:
- `name` isn't referenced outside of the class
- `setPrototypeOf` doesn't supply additional functionality